### DISCUSSION TITLE
Optimistic mutation, undo logic : Snapshot or filter?

### DISCUSSION THREAD

**Discussion by Pumkko** [TanStack/query#3901] [Q&A]
We are developing a PWA and we need our mutations to work offline, so we are using optimistic updates and `PersistQueryClientProvider`.

I have read the doc to understand how to do so with React Query and i have found those two pages
https://tanstack.com/query/v4/docs/guides/mutations#persist-mutations
https://tanstack.com/query/v4/docs/guides/optimistic-updates

The provided examples do not use the same logic to undo the mutations, the first link filters out the added from the current values, whereas the second link applies a snapshot of the values as they were before the mutation.
To make things clearer, here's some code from the links (in case they're dead)
### Using a filter
```typescript

    // In onMutate
    return { optimisticTodo }
 },
 onError: (error, variables, context) => {
    // Remove optimistic todo from the todos list
    queryClient.setQueryData(['todos'], old => old.filter(todo => todo.id !== context.optimisticTodo.id))
 },
```
### Apply a snapshot
```typescript
  // In onMutate
  // Snapshot the previous value
   const previousTodos = queryClient.getQueryData(['todos'])

    // Optimistically update to the new value, i omitted the code since it's irrelevant 

    // Return a context object with the snapshotted value
    return { previousTodos } 
 },
 onError: (err, newTodo, context) => {
  // If the mutation fails, use the context returned from onMutate to roll back
  queryClient.setQueryData(['todos'], context.previousTodos)
 },
```
Since our mutations are to be persisted for later use in case the user goes offline for days, which of those two solutions should we use in our project ? Also should we implement `onSettled` ?
```typescript
  onSettled: () => {
    queryClient.invalidateQueries(['todos'])
  },
```

**Comment by TkDodo** [comment]
I don't see where the first link you provided (https://tanstack.com/query/v4/docs/guides/mutations#consecutive-mutations) filters out anything?

I think all the examples that we provide return a `snapshot` or a `rollback` function from `onMutate` and then just apply that.

**Reply by Pumkko** [reply]
The first link was incorrect, i have edited the question, thanks for telling me.
The new link (so you don't have to scroll up) : https://tanstack.com/query/v4/docs/guides/mutations#persist-mutations

This link contains the following code

```typescript
 onError: (error, variables, context) => {
    // Remove optimistic todo from the todos list
    queryClient.setQueryData(['todos'], old => old.filter(todo => todo.id !== context.optimisticTodo.id))
  },
```

**Reply by Pumkko** [reply]
To make the things clearer, i have deleted my question from SO, so we can keep discussing this here

**Reply by TkDodo** [reply]
okay, I've never seen that example I guess. It honestly doesn't matter and depends on your use-case. maybe refetches have happened in the meantime and you snapshot is now outdated, so filtering it out is better? Also depends on if you actually refetch when an optimistic update fails or not. returning the previousData and just setting that directly is certainly less code.
